python-package/setup.py

```diff
@@ -118,8 +118,22 @@ def compile_cpp(use_mingw=False, use_gpu=False, use_mpi=False, nomp=False,
         cmake_cmd.append("-DUSE_MPI=ON")
     if nomp:
         cmake_cmd.append("-DUSE_OPENMP=OFF")
+    if system() == 'Darwin' and not nomp:
+        cc = os.environ.get('CC')
+        cxx = os.environ.get('CXX')
+        if not (cc and cc.startswith('gcc') and cxx and cxx.startswith('g++')):
+            # Apple Clang
+            # https://github.com/Homebrew/homebrew-core/pull/20589
+            cmake_cmd.extend([
```

Comment:
We shouldn't use hardcode paths for the first build attempt. Let's use hardcoded paths only after first attempt fail:
```
...
 else:  # Linux, Darwin (macOS), etc.
        logger.info("Starting to compile with CMake.")
        if system() == 'Darwin' and not nomp and not (cc and cc.startswith('gcc') and cxx and cxx.startswith('g++')):
            status = silent_call(cmake_cmd)
            if status != 0:
                logger.warning("Compilation failed.")
                logger.info("Starting to compile with OpenMP path guesses.")
                clear_path(os.path.join(CURRENT_DIR, "build_cpp"))
                <your modification here>
        else:
            silent_call(cmake_cmd, raise_error=True, error_msg='Please install CMake and all required dependencies first')
...
```

Reply:
> We shouldn't use hardcoded paths for the first build attempt.

Why is this?
I think this trial is a really waste.

Reply:
1. This workaround was tested only on High Sierra and Mojave. We don't know how it works on earlier versions. However, the current method was tested by the time and there were no any complaints.
2. There could be some other problems with this workaround which we don't know.
3. User can install OpenMP in another folder.

Reply:
To be honest, it is correct for macOS users to set cmake options for AppleClang with OpenMP.
In ohter words, the current way (`cmake ..`) is wrong (than this change) because `libomp` requires the options.

Reply:
@henry0312 @StrikerRUS What about something like this instead in README.md? (and let the user change the paths?)

```sh
DOpenMP_C_LIB_NAMES="omp" DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" DOpenMP_CXX_LIB_NAMES="omp" DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib bash -c 'cmake ..'
```

Reply:
@henry0312 
> In ohter words, the current way (cmake ..) is wrong (than this change) because libomp requires the options.

Maybe... But if everything works (OK, work**ed** before 10.14), why should we ask users do more things then needed? 

@henry0312 @Laurae2 Hmm, what do you think about new setup args `openmp-include-dir` and `openmp-library` like we already have for [GPU version](https://github.com/Microsoft/LightGBM/tree/master/python-package#build-gpu-version)?

Then we can do the following:
```
1. Try to compile with default "cmake .." for earlier macOS versions and future CMake changes.
2. If new args are provided, compile with them.
3. If not 2. (not provided or failed), then try to compile with path guesses for the maximum backward compatibility.
```

Reply:
> why should we ask users do more things then needed?

No, no.
This change will never ask users to do more anything.
For users, nothing will be changed.
For us, this change will correct building options.  That's all.